### PR TITLE
Blog Stats Block: Include Visitors Data

### LINKED ISSUE #11428: Show unique visitors with Jetpack widget
Hi

You already have this widget which shows the amount of all visitors from the beginning.
Unfortunately it shows only visitors and not unique visitors, what means that same user can count several times.

I really would appreciate to see this widget with the option to choose between visitors **and** unique visitors.

Hope to see that update in near future. :)

### PR DESCRIPTION
**Pull request by Aurorum**
Fixes #11428 

## Proposed changes:

* Adds the number of visitors to a site as something that can be displayed as part of the Blog Stats block.

## Jetpack product discussion

See original issue (technically relates to the Legacy Widget, but I think it's reasonable for the block to supersede that).

## Does this pull request change what data or activity we track or use?

We already use this data elsewhere, but it makes it visible in a new way.

## Testing instructions:

Enable Beta blocks and add the "Blog Stats" widget to your site. You should now see two radios (an option dropdown felt very overkill for just two options, so I think a radio is most appropriate here). The new one should enable you to display your site visitors instead of views. 

**Note:** we don't collect data of unique visitors for individual posts. You should be prevented from selecting the "Visitors" and "This individual post" combination.

Screenshot 2024-02-02 at 22 58 58

cc @kangzj - sorry to give you something else to review :) for what it's worth, I think this is the only significant issue for the Legacy Widget in the backlog, so would be nice to fix it with this new block!

@jeherve - I'd appreciate your advice here regarding what you said in https://github.com/Automattic/jetpack/issues/11428#issuecomment-538898604:

> I don't think that data is available in the WordPress.com REST API right now, so we'd have to extend the API first.

My understanding is that `visitors` in the API is just that, unique visitors. WordPress.com uses that term precisely from launch [here](https://wordpress.com/blog/2012/12/17/unique-visitors/) and refer to it still [here](https://wordpress.com/support/stats/understand-your-sites-traffic/#views-and-visitors). As such, I believe that the comment in https://github.com/Automattic/jetpack/issues/11428#issuecomment-538727734 was mistaken, but just wanted to make sure.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js [hunk 1]
```diff
@@ -34,6 +38,13 @@ function BlogStatsEdit( { attributes, className, setAttributes } ) {
 		}
 	}, [ postId, isModuleActive ] );
 
+	// We don't collect visitor data for individual posts.
+	useEffect( () => {
```
Comment (kangzj) [inline projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js:42]: 👍

File: projects/plugins/jetpack/extensions/blocks/blog-stats/test/controls.js [hunk 2]
```diff
@@ -19,10 +20,30 @@ describe( 'BlogStatsControls', () => {
 	} );
 
 	describe( 'Inspector settings', () => {
-		test( 'loads and displays settings', () => {
+		test( 'loads and displays views or visitors settings', () => {
```
Comment (kangzj) [inline projects/plugins/jetpack/extensions/blocks/blog-stats/test/controls.js:23]: 👍

**Review by kangzj — DISMISSED** [review]
It tests well and looks good to me! ❤️

**Comment by kangzj** [comment]
I noticed the build process is throwing an error complaining:

```
[plugins/jetpack 2:10.354] [build-production-extensions] ERROR in editor-beta.js: Translator comments have gone missing for "hit"
```

**Comment by Aurorum** [comment]
Err, it should still be there: 
https://github.com/Automattic/jetpack/blob/6e4ad75729656c6d34df4d49e810f1be981d955d/projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js#L60-L61

Any idea on what the issue might be?

File: projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php [hunk 3]
```diff
@@ -71,36 +71,39 @@ function load_assets( $attributes ) {
...
+		}
+
+		if ( $attributes['statsData'] === 'visitors' && isset( $data->stats->visitors ) ) {
+			$stats = $data->stats->visitors;
 		}
 	}
 
-	$label = $attributes['label'] ? $attributes['label'] : esc_html(
+	$fallback_label = $attributes['statsData'] === 'visitors' ? esc_html(
+		_n( 'visitor', 'visitors', $stats, 'jetpack' )
```
Comment (kangzj) [inline projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]:
```suggestion
		/* Translators: Number of visitors */
		_n( 'visitor', 'visitors', $stats, 'jetpack' )
```

Comment (kangzj) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Let's try this 🤔

**Review by kangzj — DISMISSED** [review]

File: projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php (hunk 3 above)
Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Build is still failing 🤔

Comment (kangzj) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Do you have any insights please @jeherve @anomiex? Thanks

`[plugins/jetpack 2:10.354] [build-production-extensions] ERROR in editor-beta.js: Translator comments have gone missing for "hit"`

Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Looks a bit ugly, but I think it relates to where the comment goes: https://github.com/Automattic/jetpack/pull/28351#discussion_r1071083525

Just added a commit that will maybe fix it? Not sure.

Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Was worth a try, but nope, issue seems to persist!

Comment (jeherve) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: You should be able to work around that by adding a dummy third argument, like it's done here:
https://github.com/Automattic/jetpack/blob/7693bbd1d11a9ad661dcf0c7043aa1bf65609741/projects/plugins/jetpack/extensions/blocks/map/styles.js#L22-L25

**Comment by jeherve** [comment]
> My understanding is that visitors in the API is just that, unique visitors. 

Yup, this should be good! I don't believe this was previously available for single posts in the API, but from what you're saying I'm getting that this is now available? 🥳

File: projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php (hunk 3 above)
Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Thanks, done! Not sure how to check locally, but hopefully that fixes the tests.

**Comment by Aurorum** [comment]
> I don't believe this was previously available for single posts in the API, but from what you're saying I'm getting that this is now available? 🥳

@jeherve I'm afraid that we still don't have visitor data for individual posts. I think that's because it was never collected as opposed to not being exposed in the API though.

We do have visitor data for the whole site though (which I think is what the original issue was about - the Blog Stats widget never allowed access to this!), so that's what this PR adds to the block. :) I've got around this by disabling the "Visitors" + "Individual post" combination. 

Screenshot 2024-02-05 at 12 28 39

File: projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php (hunk 3 above)
Comment (anomiex) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/blog-stats.php:92]: Part of the confusion is that this isn't the code causing that error, it's the JS code in `edit.js`. I'll comment there with details.

You should be able to check locally with `jetpack build --production plugins/jetpack`, or in `projects/plugins/jetpack/` run `pnpm build-production-extensions`.

File: projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js [hunk 4]
```diff
@@ -44,6 +55,12 @@ function BlogStatsEdit( { attributes, className, setAttributes } ) {
 		);
 	}
 
+	/* Translators: Number of visitors */
+	const visitorsPlaceholder = _n( 'visitor', 'visitors', parseInt( stats ), 'jetpack', 0 );
+
+	/* Translators: Number of views */
+	const viewsPlaceholder = _n( 'hit', 'hits', parseInt( stats ), 'jetpack', 0 );
```
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js]: The error
```
[plugins/jetpack 2:10.354] [build-production-extensions] ERROR in editor-beta.js: Translator comments have gone missing for "hit"
```
is actually coming from here. There's a detailed explanation of what's going on at https://www.npmjs.com/package/@automattic/i18n-check-webpack-plugin#lost-comments-due-to-expression-movement, but the TL;DR is that this code is getting transformed to look something like this instead:
```js
	/* Translators: Number of visitors */
	const visitorsPlaceholder = _n( 'visitor', 'visitors', parseInt( stats ), 'jetpack', 0 ),
		viewsPlaceholder = _n( 'hit', 'hits', parseInt( stats ), 'jetpack', 0 );
	/* Translators: Number of views */
```
(i.e. it deletes the first `const` and moves the second assignment into the first one, but leaves the comment behind where it was)

This seems to work
```suggestion
	const
		/* Translators: Number of visitors */
		visitorsPlaceholder = _n( 'visitor', 'visitors', parseInt( stats ), 'jetpack', 0 ),
		/* Translators: Number of views */
		viewsPlaceholder = _n( 'hit', 'hits', parseInt( stats ), 'jetpack', 0 );
```
Or this
```suggestion
	const visitorsPlaceholder =
		/* Translators: Number of visitors */
		_n( 'visitor', 'visitors', parseInt( stats ), 'jetpack', 0 );

	const viewsPlaceholder =
		/* Translators: Number of views */
		_n( 'hit', 'hits', parseInt( stats ), 'jetpack', 0 );
```

I don't know why 924d9bea didn't work. Possibly I should remove the corresponding suggestion from the doc I linked if it doesn't work anymore.

Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js]: I've moved the comment and added a parameter - hopefully that works! Thanks for taking a look. :)

Comment (kangzj) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js]:
```
[plugins/jetpack 2:10.961] [build-production-extensions] ERROR in editor-beta.js: Translator comments have gone missing for "visitor"
[plugins/jetpack 2:10.961] [build-production-extensions]  - Translators: Number of visitors
[plugins/jetpack 2:10.961] [build-production-extensions] 
[plugins/jetpack 2:10.961] [build-production-extensions] ERROR in editor-beta.js: Translator comments have gone missing for "hit"
[plugins/jetpack 2:10.961] [build-production-extensions]  - Translators: Number of views
[plugins/jetpack 2:10.961] [build-production-extensions] 
```

🤔

Comment (anomiex) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js]: Unfortunately f8a461d is basically a re-do of 924d9bea which still doesn't work (see #35452 for what I know about that). Either of the suggestions I made above worked for me locally.

Comment (Aurorum) [reply projects/plugins/jetpack/extensions/blocks/blog-stats/edit.js]: Sorry, it looks like I committed the wrong thing. 🤦 It works fine locally for me too - re-added your suggestion now!

**Review by kangzj — APPROVED** [review]
🎉

**Pull request merged**